pkg/backup/backupdest/backup_index.go

```diff
@@ -138,6 +142,169 @@ func IndexExists(ctx context.Context, store cloud.ExternalStorage, subdir string
 	return indexExists, nil
 }
 
+// ListIndexes lists all the index files for a backup chain rooted by the full
+// backup indicated by the subdir. The store should be rooted at the default
+// collection URI (the one that contains the `index/` directory). It returns
+// the basenames of the listed index files. It assumes that the subdir is
+// resolved and not `LATEST`.
+//
+// The indexes are returned in reverse chronological order, i.e. the most recent
+// index is first in the list.
+func ListIndexes(
+	ctx context.Context, store cloud.ExternalStorage, subdir string,
+) ([]string, error) {
+	var indexBasenames []string
+	if err := store.List(
+		ctx,
+		indexSubdir(subdir)+"/",
+		"",
+		func(file string) error {
+			indexBasenames = append(indexBasenames, path.Base(file))
+			return nil
+		},
+	); err != nil {
+		return nil, errors.Wrapf(err, "listing indexes in %s", subdir)
+	}
+	return indexBasenames, nil
+}
+
+// GetBackupTreeIndexMetadata concurrently retrieves the index metadata for all
+// backups within the specified subdir, up to the specified end time, inclusive.
+// It also returns the total size of memory allocated to store the index files.
+// The returned indexes are sorted in reverse chronological order, i.e. the
+// latest index is first in the list. The store should be rooted at the
+// collection URI that contains the `index/` directory.
+//
+// Note: Even if an error is returned, the total memory reserved in the monitor
+// is still returned as it may be non-zero.
+func GetBackupTreeIndexMetadata(
+	ctx context.Context,
+	mem *mon.BoundAccount,
+	store cloud.ExternalStorage,
+	subdir string,
+	endTime hlc.Timestamp,
+) ([]backuppb.BackupIndexMetadata, int64, error) {
+	indexBasenames, err := ListIndexes(ctx, store, subdir)
+	if err != nil {
+		return nil, 0, err
+	}
+	// Note that the following logic depends on the fact that ListIndexes
+	// returns the index files in reverse chronological order.
+	startIdx := 0
+	if !endTime.IsEmpty() {
+		// We will still need to open each of the index files to ensure that the
+		// backup end times (which are nano-second specific) are within the
+		// specified endTime, but we can do some preliminary filtering first based
+		// on the filename.
+		for startIdx = len(indexBasenames) - 1; startIdx >= 0; startIdx-- {
+			_, indexEnd, err := parseIndexFilename(indexBasenames[startIdx])
```

Comment:
again, i don't think filename parsing should be on the core path to backup and restore, even if it saves us a couple of seconds. I also dont think there are any memory concerns for reading 400 index metadata pbs into memory.  I think as a general rule of thumb, I would strive to:
- get the simplest (i.e. smallest diff) prototype out, without any optimizations
- then we can evaluate potential optimizations. again, i don't think this parsing of the file path will be worth our wile.

Perhaps run this by -eng or in discussion topics today, in case folks feel differently.